packages/xforms-engine/src/instance/RankControl.ts

```diff
@@ -0,0 +1,144 @@
+import type { CurrentState } from '../lib/reactivity/node-state/createCurrentState.ts';
+import type { EngineState } from '../lib/reactivity/node-state/createEngineState.ts';
+import type { SharedNodeState } from '../lib/reactivity/node-state/createSharedNodeState.ts';
+import { createSharedNodeState } from '../lib/reactivity/node-state/createSharedNodeState.ts';
+import { createFieldHint } from '../lib/reactivity/text/createFieldHint.ts';
+import { createNodeLabel } from '../lib/reactivity/text/createNodeLabel.ts';
+import type { SimpleAtomicState } from '../lib/reactivity/types.ts';
+import type { Root } from './Root.ts';
+import type { ValueNodeStateSpec } from './abstract/ValueNode.ts';
+import { ValueNode } from './abstract/ValueNode.ts';
+import type { GeneralParentNode } from './hierarchy.ts';
+import type { EvaluationContext } from './internal-api/EvaluationContext.ts';
+import type { ValidationContext } from './internal-api/ValidationContext.ts';
+import type { ClientReactiveSubmittableValueNode } from './internal-api/submission/ClientReactiveSubmittableValueNode.ts';
+import { RankFunctionalityError, RankValueTypeError } from '../error/RankError.ts';
+import { BaseItemCollectionCodec } from '../lib/codecs/BaseItemCollectionCodec.ts';
+import { sharedValueCodecs } from '../lib/codecs/getSharedValueCodec.ts';
+import type { AnyNodeDefinition } from '../parse/model/NodeDefinition.ts';
+import { createItemCollection } from '../lib/reactivity/createItemCollection.ts';
+
+type AssertRangeNodeDefinition = (definition: RankDefinition) => asserts definition is RankDefinition<'string'>;
+const assertRangeNodeDefinition: AssertRangeNodeDefinition = (definition) => {
+	if (definition.valueType !== 'string') {
+		throw new RankValueTypeError(definition);
+	}
+};
+
+type RankItemMap = ReadonlyMap<string, RankItem>;
+
+interface RankControlStateSpec extends ValueNodeStateSpec<readonly string[]> {
+	readonly label: Accessor<TextRange<'label'> | null>;
+	readonly hint: Accessor<TextRange<'hint'> | null>;
+	readonly valueOptions: Accessor<RankValueOptions>;
+}
+
+export class RankControl
+	extends ValueNode<'string', RankDefinition<'string'>, readonly string[], readonly string[]>
+	implements
+		RankNode,
+		XFormsXPathElement,
+		EvaluationContext,
+		ValidationContext,
+		ClientReactiveSubmittableValueNode
+{
+	static from(parent: GeneralParentNode, definition: RankDefinition): RankControl {
+		assertRangeNodeDefinition(definition);
+		return new this(parent, definition);
+	}
+
+	private readonly mapOptionsByValue: Accessor<RankItemMap>;
+
+	protected override readonly getInstanceValue: Accessor<string>;
+
+	// XFormsXPathElement
+	override readonly [XPathNodeKindKey] = 'element';
+
+	// InstanceNode
+	protected readonly state: SharedNodeState<RankControlStateSpec>;
+	protected readonly engineState: EngineState<RankControlStateSpec>;
+
+	// RankNode
+	readonly nodeType = 'rank';
```

Comment:
Here we'll also want to add `appearances`, see comment at the top of `RankControlDefinition.ts`.